configcatclient/rolloutevaluator.py

```diff
@@ -48,174 +49,616 @@ def evaluate(self, key, user, default_value, default_variation_id, settings):  #
-                    bucket += rollout_percentage_item.get(PERCENTAGE, 0)
-                    if hash_val < bucket:
-                        percentage_value = rollout_percentage_item.get(VALUE)
-                        variation_id = rollout_percentage_item.get(VARIATION_ID, default_variation_id)
-                        log_entries.append('Evaluating %% options. Returning %s' % percentage_value)
-                        return percentage_value, variation_id, None, rollout_percentage_item, None
-
-            return_value = setting_descriptor.get(VALUE, default_value)
-            return_variation_id = setting_descriptor.get(VARIATION_ID, default_variation_id)
-            log_entries.append('Returning %s' % return_value)
-            return return_value, return_variation_id, None, None, None
-        finally:
-            self.log.info('%s', '\n'.join(log_entries), event_id=5000)
+        # Circular dependency check
+        visited_keys = context.visited_keys
+        visited_keys.append(context.key)
+        if prerequisite_key in visited_keys:
+            depending_flags = ' -> '.join("'{}'".format(s) for s in list(visited_keys) + [prerequisite_key])
+            error = 'Cannot evaluate condition (%s) for setting \'%s\' (circular dependency detected between the following ' \
+                    'depending flags: %s). Please check your feature flag definition and eliminate the circular dependency.'
+            error_args = (prerequisite_condition, context.key, depending_flags)
+            self.log.warning(error, *error_args, event_id=3005)
+            if log_builder:
+                log_builder.append(prerequisite_condition + ' ')
+
+            if visited_keys:
+                visited_keys.pop()
+
+            return prerequisite_condition_result, 'cannot evaluate, circular dependency detected'
+
+        if log_builder:
+            log_builder.append(prerequisite_condition)
+            log_builder.new_line('(').increase_indent()
+            log_builder.new_line("Evaluating prerequisite flag '%s':" % prerequisite_key)
+
+        prerequisite_value, _, _, _, error = self.evaluate(prerequisite_key, context.user, None, None, config,
+                                                           log_builder, context.visited_keys)
+
+        if visited_keys:
+            visited_keys.pop()
+
+        if error is not None:
+            return prerequisite_condition_result, error
+
+        if log_builder:
+            log_builder.new_line("Prerequisite flag evaluation result: '%s'." % str(prerequisite_value))
+            log_builder.new_line("Condition (Flag '%s' %s '%s') evaluates to " %
+                                 (prerequisite_key, PREREQUISITE_COMPARATOR_TEXTS[prerequisite_comparator],
+                                  str(prerequisite_comparison_value)))
+
+        # EQUALS
+        if prerequisite_comparator == PrerequisiteComparator.EQUALS:
+            if prerequisite_value == prerequisite_comparison_value:
```

Comment:
This may require further discussion because we haven't really defined the behavior in the case of type mismatch. It pretty much looks like a "cannot evaluate" case to me.